package NYT.NQueryClient.NProto;

import "yt/ytlib/chunk_client/chunk_spec.proto";
import "yt/ytlib/query_client/expression.proto";

////////////////////////////////////////////////////////////////////////////////

message TOperator
{
    required int32 kind = 1;

    extensions 100 to max;
}

message TScanOperator
{
    extend TOperator {
        optional TScanOperator scan_operator = 101;
    }
    required int32 table_index = 1;
    required NYT.NChunkClient.NProto.TChunkSpec data_split = 2;
}

message TUnionOperator
{
    extend TOperator {
        optional TUnionOperator union_operator = 102;
    }
    repeated TOperator sources = 1;
}

message TFilterOperator
{
    extend TOperator {
        optional TFilterOperator filter_operator = 103;
    }
    required TOperator source = 1;
    required TExpression predicate = 2;
}

message TProjectOperator
{
    extend TOperator {
        optional TProjectOperator project_operator = 104;
    }
    required TOperator source = 1;
    repeated TExpression projections = 2;
}

////////////////////////////////////////////////////////////////////////////////

